### PR TITLE
feat: adding an auto completion component in chat playground

### LINKED ISSUE #547: Adding an auto-completion component for /playground for local function / files
With the new v1beta/search API (nightly release), we can generate auto-completion when a user is attempting to reference an identifier in the chat box (e.g., function name, class name).

This auto-completion should trigger whenever the user types an "@".

Code location: [here](https://github.com/TabbyML/tabby/blob/main/ee/tabby-ui/components/prompt-form.tsx#L22)

### PR DESCRIPTION
**Pull request by liangfung**
Close #547

### CODE REVIEW

File: ee/tabby-ui/components/prompt-form.tsx [hunk 1]
```diff
@@ -26,15 +35,56 @@ export function PromptForm({
...
+  const [options, setOptions] = React.useState<SearchReponse['hits']>([])
+  const onSearch = React.useMemo(() => {
+    return debounce((e: React.ChangeEvent<HTMLTextAreaElement>) => {
+      const value = e.target?.value ?? ''
+      const end = e.target?.selectionEnd ?? 0
+      const queryname = getSearchCompletionQueryName(value, end)
+      if (queryname) {
+        let url = `/v1beta/search?q=name:${queryname}`
+        latestFetchKey.current = url
+        setQueryCompletionUrl(`/v1beta/search?q=name:${queryname}`)
```
Comment (wsxiaoys) [inline ee/tabby-ui/components/prompt-form.tsx:62]:
```suggestion
        const query = encodeURIComponent(`name:${queryname} kind:function`)
        const url = `/v1beta/search?q=${query}`
        latestFetchKey.current = url
        setQueryCompletionUrl(url)
```

File: ee/tabby-ui/components/prompt-form.tsx [hunk 2]
```diff
@@ -26,15 +35,56 @@ export function PromptForm({
...
+    }, 200)
   }, [])
 
+  const onSelectCompletion = (
+    inputRef: React.MutableRefObject<
+      HTMLTextAreaElement | HTMLInputElement | null
+    >,
+    item: ISearchHit
+  ) => {
+    const replaceString = '`@' + item?.doc?.name + '` '
```
Comment (wsxiaoys) [inline ee/tabby-ui/components/prompt-form.tsx:75]: no need to replace @get_name, just keep it as is.

File: ee/tabby-ui/components/prompt-form.tsx [hunk 3]
```diff
@@ -26,15 +35,56 @@ export function PromptForm({
...
+    if (queryname) {
+      setInput(prevInput + replaceString + input.substring(selectionEnd))
+    }
+
+    setOptions([])
+  }
+
   return (
     <form
       onSubmit={async e => {
```
Comment (wsxiaoys) [inline ee/tabby-ui/components/prompt-form.tsx:90]: When submit, @get_name needs to be replaced with `doc.body`

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-ui/lib/utils.ts [hunk 4]
```diff
@@ -41,3 +41,18 @@ export function formatDate(input: string | number | Date): string {
     year: 'numeric'
   })
 }
+
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/utils.ts:44]: should be good to leave this function in `prompt-form.tsx` until there's a need for reuse.

File: ee/tabby-ui/package.json [hunk 5]
```diff
@@ -30,7 +31,9 @@
     "class-variance-authority": "^0.4.0",
     "clsx": "^1.2.1",
     "compare-versions": "^6.1.0",
+    "downshift": "^8.2.2",
     "focus-trap-react": "^10.1.1",
+    "lodash": "^4.17.21",
```
Comment (wsxiaoys) [inline ee/tabby-ui/package.json:36]: use `lodash-es` for tree shaking?

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

**Comment by wsxiaoys** [comment]
UI review: https://www.figma.com/file/j9O1Z5LPbXlYNhF8fCmcNN/Tabby-AutoComplete-Component?type=design&node-id=0%3A1&mode=design&t=94NDryO1POHUVPXf-1

File: ee/tabby-ui/components/prompt-form.tsx (hunk 2 above)
Comment (liangfung) [reply ee/tabby-ui/components/prompt-form.tsx:75]: keep it as is and wrap with `` ?

Comment (liangfung) [reply ee/tabby-ui/components/prompt-form.tsx:75]: keep it as is and wrap with `` ?

Comment (wsxiaoys) [reply ee/tabby-ui/components/prompt-form.tsx:76]: i'll so it's not necessary to wrap with ``. But certain highlighting on `@get_function` will be somewhat helpful (not required in this PR)

Comment (liangfung) [reply ee/tabby-ui/components/prompt-form.tsx]: Done. And a space is inserted after the input to break the query string.

File: ee/tabby-ui/components/prompt-form.tsx (hunk 3 above)
Comment (liangfung) [reply ee/tabby-ui/components/prompt-form.tsx]: Done

File: ee/tabby-ui/package.json (hunk 5 above)
Comment (liangfung) [reply ee/tabby-ui/package.json]: Great suggestion, done

**Comment by liangfung** [comment]
> UI review: https://www.figma.com/file/j9O1Z5LPbXlYNhF8fCmcNN/Tabby-AutoComplete-Component?type=design&node-id=0%3A1&mode=design&t=94NDryO1POHUVPXf-1

Please check the UI and let me know if more modifications are needed

**Comment by wsxiaoys** [comment]
Some UI findings:
https://jam.dev/c/baefcf2e-852c-4de5-bd90-8c7205214d22
https://jam.dev/c/54d59d9c-5273-4d48-9e20-e59b52939039

**Review by wsxiaoys — APPROVED** [review]

**Pull request merged**